### PR TITLE
analytics: generate a cohort ID based on user's first week and log it automatically with all events

### PR DESCRIPTION
**Pull request by limitedmage**
Cohort ID is generated at the same time the Anonymous User ID is generated and is the date of the Monday of the week the user first visited. Just like the Anonymous User ID, the cohort ID is saved in a cookie and is logged with every call to `logEvent`.

Users that already have an Anonymous User ID before this PR is merged won't have a Cohort ID.

### CODE REVIEW

**Review by camdencheek — APPROVED** [review]

File: client/web/src/tracking/eventLogger.ts [hunk 1]
```diff
@@ -1,21 +1,35 @@
...
 
 export class EventLogger implements TelemetryService {
     private hasStrippedQueryParameters = false
 
-    private anonymousUserId?: string
+    private anonymousUserID?: string
+    private cohortID?: string
     private firstSourceURL?: string
 
+    private readonly CookieSettings: CookieAttributes = {
```
Comment (novoselrok) [inline client/web/src/tracking/eventLogger.ts:21]: This can be lowercase, `cookieSettings`.

File: client/web/src/tracking/eventLogger.ts [hunk 2]
```diff
@@ -71,29 +85,39 @@ export class EventLogger implements TelemetryService {
      * Get the anonymous identifier for this user (used to allow site admins
      * on a Sourcegraph instance to see a count of unique users on a daily,
      * weekly, and monthly basis).
+     * If the user doesn't have an anonymours ID yet, it generates one as well
```
Comment (novoselrok) [inline client/web/src/tracking/eventLogger.ts:88]: `anonymours` -> `anonymous`

File: client/web/src/user/settings/backend.tsx [hunk 3]
```diff
@@ -153,6 +155,7 @@ export function logEvent(event: string, eventProperties?: unknown): void {
         {
             event,
             userCookieID: eventLogger.getAnonymousUserID(),
+            cohortID: eventLogger.getCohortID() || null, // Must be called AFTER eventLogger.getAnonymousUserID()
```
Comment (novoselrok) [inline client/web/src/user/settings/backend.tsx:158]: Thought: could we make it that `getCohortID` and `getAnonymousUserID` are independent, so it doesn't matter in what order they're called? 

We could generate the `anonymousUserID` and `cohortID` in the `EventLogger` constructor instead, as well as set the cookies. If you feel that's too big of a change, then I'm fine with the current solution.

File: client/web/src/tracking/util.ts [hunk 4]
```diff
@@ -34,3 +36,23 @@ export function redactSensitiveInfoFromURL(url: string): string {
 
     return sourceURL.href
 }
+
+/**
+ * Returns the Monday at or before the supplied date, in YYYY-MM-DD format.
+ * This is used to generate cohort IDs for users who
+ * started using the site on the same week.
+ */
+export function getPreviousMonday(date: Date): string {
```
Comment (novoselrok) [inline client/web/src/tracking/util.ts:45]: Here is a one-liner that should also work for this (using `date-fns`): `formatISO(startOfWeek(new Date(), { weekStartsOn: 1 }), { representation: 'date' })` 😄

**Review by novoselrok — APPROVED** [review]

**Review by malomarrec — APPROVED** [review]

File: client/web/src/user/settings/backend.tsx (hunk 3 above)
Comment (limitedmage) [reply client/web/src/user/settings/backend.tsx:158]: I like this much better, I'll change it to do this.

File: client/web/src/tracking/util.ts (hunk 4 above)
Comment (limitedmage) [reply client/web/src/tracking/util.ts:45]: I forgot we had `date-fns`, thank you!! I spent way too much time on this function >_>

**Pull request merged**